### ISSUE TITLE
Replica change operations need additional transactional safety

### ISSUE THREAD

**Issue by mrtracy** [cockroachdb/cockroach#2152]
The various replica change operations (AdminSplit, AdminMerge, ChangeReplica) are not transactionally safe for their current usage.

Each of these methods attempt to maintain transactional safety with this pattern:
```go
    // Lock a local 'metaLock' on the replica.
    r.metaLock.Lock()
    defer r.metaLock.Unlock()

    // Get the current value of the descriptor
    desc := r.Desc()

    // Prepare some updates to the descriptor...

    // Open a transaction and Cput the descriptor. 
    err := r.rm.DB().Txn(func(txn *client.Txn) error {
        b := &client.Batch{}
        descKey := keys.RangeDescriptorKey(updatedDesc.StartKey)

        // updateRangeDescriptor performs a CPut on the range descriptor.
        if err := updateRangeDescriptor(b, descKey, desc, &updatedDesc); err != nil {
            return err
        }
        //...
    }
```

The Cput inside of `updateRangeDescriptor` will fail the transaction if changes have been made to the range descriptor by another node process since `r.Desc()` was read.

However, `r.Desc()` was read within the method, which is *after* some external method has already decided to initiate a split, merge or change.  This creates the following race:

+ Two replicas of the same range decide that it is appropriate to split.
+ Both replicas kick off an AdminSplit() call concurrently.
+ The execution of AdminSplit() on the first node finishes before the AdminSplit() on the second begins.  As a result, both calls will succeed and the range will be split twice.

This appears to create the risk of double splits, or in the case of rebalance double adds / double removes.

To fix this, `r.Desc()` must be called before a decision is made whether to call one of these methods.  If the decision is positive, that descriptor must be passed to the method, rather than having the method call `r.Desc()` itself.

#### Note on Replica.metaLock
The current code attempts to solve this issue by using the local `Replica.metaLock` mutex.  For splits and merges (but not ChangeReplicas), it is enforced that AdminSplit or AdminMerge can only be executed on the current range leader, and the range leading replica can only be executing one such operation at a time because of `metaLock`.

However, there appears to be no guarantee that a replica will remain the range leader through the execution of the Admin command; it's possible, although unlikely, for another replica to become the leader and execute the same admin command.

Therefore, `metaLock` appears to be offering no real race protection.  Instead, we should rely on the CPut of the range descriptor to prevent replica change races.  metaLock itself should be removed, because it gives the illusion of safety that is not present.

**Comment by tamird** [comment]
This looks like a mini-RFC, so: I approve :+1:

**Comment by bdarnell** [comment]
SGTM

**Comment by tschottdorf** [comment]
The plan is to remove `metaLock` and enforce in the `Txn` that the descriptor active during the `Txn` is the one we've read? LGTM

**Comment by mrtracy** [comment]
This was fixed by #2241
